File: events/star.go
package events

import (
	"fmt"
	"time"
	"webhook/config"
	"webhook/context"
	"webhook/debouncer"
	"webhook/structs/discord"
	"webhook/structs/github"

	"github.com/rs/zerolog/log"
)

type star struct {
	Action     string            `json:"action"`
	Sender     github.User       `json:"sender"`
	Repository github.Repository `json:"repository"`
}

func (e *star) handle(*context.Context) (*discord.Webhook, error) {
	if e.Action != "created" {
		return nil, nil
	}

	ok := debouncer.Debounce("star", e.Sender.Login, e.Repository.Name, time.Hour*24)
	if !ok {
		log.Debug().Str("repository", e.Repository.Name).Str("username", e.Sender.Login).Msg("Event is currently debounced")
		return nil, nil
	}

	return &discord.Webhook{
		Content: fmt.Sprintf(
			"[%s](<%s>) starred [%s](<%s>) %s",
			e.Sender.Login,
			e.Sender.HtmlUrl,
			e.Repository.Name,
			e.Repository.HtmlUrl,
			config.Get().HappyEmoji,
		),
		Username:  e.Sender.Login,
		AvatarUrl: e.Sender.AvatarUrl,
	}, nil
}

File: src/events/star.rs
use crate::errors::Error;
use crate::events::Event;
use crate::events::base::{Repository, User, WebhookMessage};
use serde::Deserialize;

#[derive(Deserialize)]
pub struct StarEvent {
    pub action: String,
    pub sender: User,
    pub repository: Repository,
}

impl Event for StarEvent {
    fn handle(&self) -> Result<Option<WebhookMessage>, Error> {
        if self.action != "created" {
            return Ok(None);
        }

        Ok(Some(WebhookMessage {
            content: format!(
                "[{}](<{}>) starred [{}](<{}>) <:foxtada:1311327105300172882>",
                self.sender.login,
                self.sender.html_url,
                self.repository.name,
                self.repository.html_url
            ),
            username: self.sender.login.clone(),
            avatar_url: self.sender.avatar_url.clone(),
        }))
    }
}
